Struct bit-fields fix

shmarovfedor opened the pull request:
It looks like currently ESBMC is struggling with structures featuring bit-fields. A simple example of this is issue #451. The main problem is that in the current memory model ESBMC operates with bytes (e.g., calculating offsets, object sizes, etc.). Hence, the problem arises when a bit-field is not at the end of the struct.

This is my first attempt at trying to tackle this issue. Briefly, I've created a method for computing pointer offset in bits (by reworking the existing method for computing pointer offset in bytes). Then I've added the code (please note it's just a draft of the logic behind the solution) for dereferencing bitfields in statically and dynamically allocated structures. While the former seems quite straightforward, I've experienced some difficulties with the latter. While I managed to extract the data correctly through a series of bit shifts, this approach fails when we try to assign a value directly to the bit-field. (Also when dealing with the signed values some further work needs to be done.) It looks like I will need to implement methods for bit extracts and bit assignments (i.e., when we need to modify only a few bits within a byte).

On the other hand, a possibly more robust solution would be modifying the existing memory model to be able to reason in bits. This however, will involve quite a lot of changes which might not be worth it considering we need to deal with bits only in a handful of cases.

Meanwhile I will continue working on this PR. Please let me know what you think about this direction or if there are better paths to be explored for tackling this issue.

shmarovfedor committed:
added pointer offset computation in bits

shmarovfedor committed:
first draft for statically allocated structs

shmarovfedor committed:
added a solution for dynamically allocated structs with unsigned bitfields

shmarovfedor committed:
added testcases for struct bitfields

mikhailramalho commented:
> On the other hand, a possibly more robust solution would be modifying the existing memory model to be able to reason in bits. This however, will involve quite a lot of changes which might not be worth it considering we need to deal with bits only in a handful of cases.
> 
> Meanwhile I will continue working on this PR. Please let me know what you think about this direction or if there are better paths to be explored for tackling this issue.

I think moving forward is best if we use bits everywhere, mostly to avoid different logics bundled together in the same memory model: that's actually the case today, where I've added some code to handle bit-fields in a memory model that only understand bytes, and that's been a source of a lot of issues ever since.

It is indeed a lot of work, including changes to our SMT backend (the model of pointer_offiset, and others), but thankfully we have a lot of test cases available, so you could try some TDD.

Also, I think some of the little-endian/big-endian code is wrong, so bear that in mind when working on the memory model.

shmarovfedor committed:
Merge branch 'master' into bitfields-fix

shmarovfedor committed:
added experimental bit-precision option

shmarovfedor committed:
added experimental bit-precision option

shmarovfedor committed:
changed malloc size to bits

shmarovfedor committed:
temporarily changed domain width calculation in smt

mikhailramalho commented:
Any update here, @shmarovfedor? It would be great if we could finish this for SV-COMP 2022.

shmarovfedor committed:
started optional type size calculation in bits

shmarovfedor commented:
@mikhailramalho , I'm working on moving all offset and type size calculations in bits atm. There are a lot of places in the code (e.g., alignment calculation, structure padding, etc.) that has a mixture of size calculation in bytes (which is then multiplied by 8) and in bits. I've started doing this for arrays (seems to be working but the backend does not support bits yet).

So I think that the next steps are: continue moving everything to bit offsets and sizes, and develop methods for representing bit extracts (probably something similar to byte extract, which seems to be the main/only way of accessing dynamically allocated memory within ESBMC memory model) and encoding them into smt. I will work on these simultaneously. Please let me know if you think it's the right direction to follow.

shmarovfedor committed:
Merge branch 'master' into bitfields-fix

rafaelsamenezes commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -29,6 +29,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/std_expr.h>
 #include <util/type_byte_size.h>
 
+#include <iostream>
```
Avoid using iostream

lucasccordeiro replied:
I agree with @rafaelsamenezes here.

rafaelsamenezes commented on src/goto-symex/symex_stack.cpp:
```diff
@@ -36,5 +43,12 @@ void goto_symex_statet::framet::decrease_stack_frame_size(const expr2tc &expr)
   // total number of bits for a given stack frame.
-  stack_frame_total -= (type_byte_size(decl_code.type) * 8);
+  if(config.options.get_bool_option("use-bit-precision"))
+  {
```
Remove `{` for single instruction ifs/else

rafaelsamenezes commented on src/goto-symex/symex_stack.cpp:
```diff
@@ -36,5 +43,12 @@ void goto_symex_statet::framet::decrease_stack_frame_size(const expr2tc &expr)
+    stack_frame_total -= type_byte_size(decl_code.type);
+  }
+  else
+  {
```
Remove `{` for single instruction ifs/else

rafaelsamenezes commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -177,24 +183,26 @@ void dereferencet::dereference_guard_expr(
-        guard.add(tmp);
-      }
-      else
-      {
```
Remove `{` for single instruction ifs/else

rafaelsamenezes commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -177,24 +183,26 @@ void dereferencet::dereference_guard_expr(
+          not2tc tmp(op);
+          guard.add(tmp);
+        }
+        else
```
Remove `{` for single instruction ifs/else

rafaelsamenezes commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -345,6 +353,89 @@ expr2tc dereferencet::dereference_expr_nonscalar(
+            //type2tc new_type(new unsignedbv_type2t(num_bytes * 8));
+            type2tc new_type(new signedbv_type2t(num_bytes * 8));
+            stitch_together_from_byte_array(value, new_type, bytes);
+            delete[] bytes;
```
expr2tc is a shared_ptr, is this needed?

rafaelsamenezes commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -1176,6 +1270,117 @@ void dereferencet::construct_from_const_struct_offset(
+    }
+
+    /*
+    // Fedor: probably cannot happen when dealing with bitfields
```
Remove comment block

rafaelsamenezes reviewed (changes requested):
Great progress so far. Just some code-style issues.

shmarovfedor committed:
modified dereferencing for bitfields allocated dynamically

shmarovfedor committed:
added return value to the function and removed some comments

mikhailramalho commented:
> @mikhailramalho , I'm working on moving all offset and type size calculations in bits atm. There are a lot of places in the code (e.g., alignment calculation, structure padding, etc.) that has a mixture of size calculation in bytes (which is then multiplied by 8) and in bits. I've started doing this for arrays (seems to be working but the backend does not support bits yet).
> 
> So I think that the next steps are: continue moving everything to bit offsets and sizes, and develop methods for representing bit extracts (probably something similar to byte extract, which seems to be the main/only way of accessing dynamically allocated memory within ESBMC memory model) and encoding them into smt. I will work on these simultaneously. Please let me know if you think it's the right direction to follow.

It think it's the way to go, hopefully in the future we can even drop the extract/update byte irep, and have only bit updates.

shmarovfedor committed:
[dereference] transformed offset calculations to bits from dereference_nonscalar_expr to build_reference_rec

shmarovfedor committed:
[dereference] converted all offsets and type sizes to bits in the recursive methods (draft)

shmarovfedor committed:
[dereference] moved offset and size types in bits to byte extraction methods

shmarovfedor committed:
[dereference] removed construct_from_bit_struct_offset method as unnecessary

shmarovfedor committed:
[dereference] did some code clean up

shmarovfedor committed:
[dereference] added bit-extraction methods

shmarovfedor committed:
[symex] added methods for handling assignments to bit-fields

shmarovfedor committed:
[regression] changed KNOWNBUG to CORE for Float22 which has been fixed now

shmarovfedor committed:
[dereference] removed old solution for bitfields

shmarovfedor committed:
[smt] reverted domain width calculation changes

shmarovfedor committed:
[options] removed use-bit-precision option

shmarovfedor committed:
[dereference] removed use-bit-precision option

shmarovfedor committed:
[util] removed use-bit-precision option

shmarovfedor committed:
[symex] removed use-bit-precision option

shmarovfedor committed:
[goto-programs] did some clean up

shmarovfedor committed:
Merge branch 'master' into bitfields-fix

shmarovfedor committed:
Merge branch 'bitfields-fix' of github.com:shmarovfedor/esbmc into bitfields-fix

shmarovfedor committed:
[code-style] applied clang-format

shmarovfedor committed:
[symex] did a clean-up and updated code formatting

shmarovfedor committed:
[dereference] fixed undeclared identifier

shmarovfedor committed:
Merge branch 'master' into bitfields-fix

shmarovfedor committed:
[dereference] changed the signature of stitch_together_from_bytes(

shmarovfedor committed:
[dereference] fixed bad_cast error

lucasccordeiro commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -661,15 +665,16 @@ expr2tc dereferencet::build_reference_to(
     return expr2tc();
 
+  // Value set tracking emits objects with some cruft built on top of them.
+  value = get_base_object(value);
```
Shall we also consider the offset?

shmarovfedor replied:
I'm not sure I fully understand the question. The final offset is computed as a sum of two offsets: the offset (i.e., "o.offset") to the object (e.g., array, struct, union, ...) plus the offset within the object (i.e., "lexical_offset") that is added later on in this method. Is there anything else that we need to take into account?

lucasccordeiro replied:
No. I believe that's all. The question was to make sure that we consider both offsets.

lucasccordeiro commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -1768,10 +1839,22 @@ expr2tc *dereferencet::extract_bytes_from_scalar(
   }
 
+  // Get the source_object and update the index. This is so that
+  // the extraction provess does not get stuck on the same index.
```
"provess" -> "process".

lucasccordeiro commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -2087,9 +2166,15 @@ void dereferencet::check_data_obj_access(
 {
+  // Perform conversion to bytes here until we figure out
+  // how to check alignment for bitfields.
+  minwidth = (minwidth + 7) / 8;
```
Shall we explain the magic numbers (e.g., why 7 )?

shmarovfedor replied:
I've added an explanation now

lucasccordeiro commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -2110,3 +2195,74 @@ void dereferencet::check_alignment(
+    unsigned long offset_int = to_constant_int2t(offset).value.to_uint64();
+    unsigned int bits_in_first_byte = offset_int % 8;
+    num_bytes =
+      (bits_in_first_byte + num_bits + 7) / 8 - (bits_in_first_byte / 8);
```
Explain why 7.

shmarovfedor replied:
Also added an explanation now

lucasccordeiro commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -2110,3 +2195,74 @@ void dereferencet::check_alignment(
+  {
+    // If everything is aligned to 8 bits, we can just return the initial value
+    if(
+      num_bits % 8 == 0 && to_constant_int2t(offset).value.to_uint64() % 8 == 0)
```
Shall we add parenthesis here to make the computation order clear?

lucasccordeiro reviewed (changes requested):
Here are some minor comments.

Could you also add more regression tests, particularly from SV-COMP?

shmarovfedor committed:
[dereference] added some comments and tidied up the code

shmarovfedor committed:
[regression] added testcases with bitfields from SV-COMP

lucasccordeiro reviewed (approved):
Overall, this PR looks good to me.

Note, however, that we must run this PR over SV-COMP before merging it into the master branch.

mikhailramalho commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -658,6 +662,56 @@ void goto_symext::symex_assign_extract(
+
+  // Simplify both expression before the recursive call
+  simplify(new_rhs);
+  simplify(value_expr);
```
Do you need this here? I think both will be simplified in `symex_assing_symbol`.

mikhailramalho reviewed (commented):
Yeah, we need a comparison of this patch against SV-COMP's benchmark to make sure we are not regressing anything.

I do have a few questions tho: 
1. Do we still need `compute_pointer_offset`? Can't we just use `compute_pointer_offset_bits`?
2. How does the cex look for the new `symex_assign_bitfield`?

fbrausse commented:
Are the following examples in the scope of this PR?

test5.c
```
#include <stdio.h>
#include <stdint.h>
#include <limits.h>
#include <assert.h>
#include <string.h>

void print_bits(uint32_t k)
{
	size_t n = sizeof(k)*CHAR_BIT;
	for (size_t i=0; i<n;) {
		printf("%c", (k >> (n - i - 1)) & 1 ? '1' : '0');
		if (++i % CHAR_BIT == 0)
			printf(" ");
	}
	printf("= %u\n", k);
}

int main(int argc, char **argv)
{
	struct {
		unsigned int s : 5;
		unsigned short f : 7;
		unsigned char : 2;
		unsigned int c : 3;
	} v = { .c = 1 << 2 | 1, .s = 1 << 4 | 1, .f = 1 << 6 | 1 };
	uint32_t i;
	memcpy(&i, &v, sizeof(i));
#ifndef __clang__
	print_bits(i);
#endif
	assert(i == 84017);
}
```
leading to
```
../build/src/esbmc/esbmc ~/test5.c 
ESBMC version 6.8.0 64-bit x86_64 linux
file /home/kane/test5.c: Parsing
Converting
Generating GOTO Program
GOTO program creation time: 0.079s
GOTO program processing time: 0.001s
Starting Bounded Model Checking
Unwinding loop 2  iteration  1   file string.c line 263 function memcpy
Unwinding loop 2  iteration  2   file string.c line 263 function memcpy
Unwinding loop 2  iteration  3   file string.c line 263 function memcpy
Unwinding loop 2  iteration  4   file string.c line 263 function memcpy
Symex completed in: 0.002s (32 assignments)
Slicing time: 0.000s (removed 25 assignments)
Generated 17 VCC(s), 1 remaining after simplification (7 assignments)
No solver specified; defaulting to z3
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
Encoding to solver time: 0.000s
Solving with solver Z3 v4.8.11
Encoding to solver time: 0.000s
Runtime decision procedure: 0.001s
Building error trace

Counterexample:

State 3 file assert.h line 109 function main thread 0
----------------------------------------------------
Violated property:
  file assert.h line 109 function main
  assertion i == 84017
  i == 84017


VERIFICATION FAILED
```

and the very similar

test6.c
```
#include <stdio.h>
#include <stdint.h>
#include <limits.h>
#include <assert.h>

void print_bits(uint32_t k)
{
	size_t n = sizeof(k)*CHAR_BIT;
	for (size_t i=0; i<n;) {
		printf("%c", (k >> (n - i - 1)) & 1 ? '1' : '0');
		if (++i % CHAR_BIT == 0)
			printf(" ");
	}
	printf("= %u\n", k);
}

int main(int argc, char **argv)
{
	union {
		struct {
			unsigned int s : 5;
			unsigned short f : 7;
			unsigned char : 2;
			unsigned int c : 3;
		};
		uint32_t i;
	} v = { .c = 1 << 2 | 1, .s = 1 << 4 | 1, .f = 1 << 6 | 1 };
#ifndef __clang__
	print_bits(v.i);
#endif
	assert(v.i == 84017);
}
```
leading to
```
../build/src/esbmc/esbmc ~/test6.c 
ESBMC version 6.8.0 64-bit x86_64 linux
file /home/kane/test6.c: Parsing
Converting
Generating GOTO Program
GOTO program creation time: 0.057s
GOTO program processing time: 0.000s
Starting Bounded Model Checking
Symex completed in: 0.000s (17 assignments)
Slicing time: 0.000s (removed 13 assignments)
Generated 2 VCC(s), 1 remaining after simplification (4 assignments)
No solver specified; defaulting to z3
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
terminate called after throwing an instance of 'z3::exception'
  what():  Sorts (Array (_ BitVec 3) (_ BitVec 8)) and (Array (_ BitVec 2) (_ BitVec 8)) are incompatible
Aborted
```

rafaelsamenezes commented:
I have pushed a version into sv-comp22-fedor for testing

lucasccordeiro commented:
@fbrausse: I believe so. 

@mikhailramalho has also a valid argument about the counterexample generation.

@shmarovfedor: could you please take a look at @fbrausse and @mikhailramalho comments?

shmarovfedor committed:
[std_types] added is_unnamed_bitfield property to componentt class

shmarovfedor committed:
[clang-c-frontend] fixed initialization for structs and unions with unnamed bitfields

shmarovfedor committed:
[dereference] added alignment check skip for bitfields

shmarovfedor committed:
[code-style] fixed code styling

shmarovfedor committed:
[symex] removed unnecessary simplifications in symex_assign_bitfield

shmarovfedor committed:
[regression] added a testcase for initialization of structs with bitfields

shmarovfedor committed:
[clang-c-frontend] fixed padding for unions

shmarovfedor committed:
Merge branch 'master' into bitfields-fix

shmarovfedor committed:
Merge branch 'bitfields-fix' of github.com:shmarovfedor/esbmc into bitfields-fix

lucasccordeiro commented:
@shmarovfedor: Does your PR support this example?

```C
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>

typedef struct str1_t {
    unsigned int field11 : 16;
    unsigned int field12 : 8;
    uint8_t field13[59];
} str1_t;


typedef struct str2_t {
    str1_t field1;
    uint32_t field2;
    uint64_t field3;
    uint32_t field4;
    uint32_t field5[4];
    uint32_t field6;
} str2_t;

uint64_t global_data[100];

int main(void) {

    str2_t var_data;
    str2_t *var_data_p = &var_data;
    uint64_t var_offset[100]; // = global_data; //= 0xdeadbeef;
    
    printf("size=%d\n", sizeof(str1_t)/sizeof(uint64_t));

    str1_t  tmp;
    uint64_t *uint_p = (uint64_t *)&tmp;
    
    uint_p[0] = 5; // ESBMC fails on this line.
    
    uint_p[1] = 5;
 
    for (uint16_t i=1; i < sizeof(str1_t)/sizeof(uint64_t); i++) { // if the for statement starts fom zero ESBMC fails.
        //((uint64_t *)&var_data_p->field1)[i] = 5; // ((uint64_t *)var_offset)[i];
        uint_p[i] = 5;
    }
   return 0; 
}
```

shmarovfedor committed:
[regression] added an example of cex generation for bitfields

shmarovfedor commented:
@mikhailramalho , I've removed the unnecessary simplifications in the `symex_assign_bitfield` now. As for `compute_pointer_offset` we still need to keep it as it's used elsewhere in the code. 

Regarding the counter-examples, I've added a testcase (`esbmc/regression/esbmc/struct_bitfields_15/`) which produces the following counter-example:
```
Counterexample:

State 3 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 | 0 & -2

State 4 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 | 0 & -3

State 5 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 12 | 0 & -13

State 6 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 | CONCAT(byte_extract_little_endian({ 12, 0, 0, 0, 0, 0, 0, 0 }, 1), byte_extract_little_endian({ 12, 0, 0, 0, 0, 0, 0, 0 }, 0)) & 65039

State 7 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 | CONCAT(byte_extract_little_endian({ 12, 0, 0, 0, 0, 0, 0, 0 }, 1), byte_extract_little_endian({ 12, 0, 0, 0, 0, 0, 0, 0 }, 0)) & 65039

State 8 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 (00000000 00000000 00000000 00000000)

State 9 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 (00000000 00000000 00000000 00000000)

State 10 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 (00000000 00000000 00000000 00000000)

State 11 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 (00000000 00000000 00000000 00000000)

State 12 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 (00000000 00000000 00000000)

State 13 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 (00000000 00000000 00000000)

State 14 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 (00000000 00000000 00000000)

State 17 file assert.h line 99 function main thread 0
----------------------------------------------------
Violated property:
  file assert.h line 99 function main
  assertion main
  !((_Bool)((signed long int)(!((signed int)((unsigned int)ptr->a) > 3))))


VERIFICATION FAILED
```

@fbrausse , thanks for pointing out these issues. As for the first one (test5.c), this was a problem with struct initialisation in the frontend where ESBMC did not correctly account for unnamed bit-fields. This has now been fixed in this PR, and I added your example to the regressions. Also upon further investigation I discovered another issue (#552) where ESBMC does not allow multiple unnamed bitfields within a struct.

Regarding your second example (test6.c) I haven't had a chance to fix it yet, but I think this problem occurs because unions flatenning does not account for bitfields in unions correctly. I will be further investigating this issue.

I'm all set up to run this PR over SV-COMP now, and I'll post the results asap.

shmarovfedor commented:
> @shmarovfedor: Does your PR support this example?
> 
> ```c
> #include <stdbool.h>
> #include <stdint.h>
> #include <stdio.h>
> 
> typedef struct str1_t {
>     unsigned int field11 : 16;
>     unsigned int field12 : 8;
>     uint8_t field13[59];
> } str1_t;
> 
> 
> typedef struct str2_t {
>     str1_t field1;
>     uint32_t field2;
>     uint64_t field3;
>     uint32_t field4;
>     uint32_t field5[4];
>     uint32_t field6;
> } str2_t;
> 
> uint64_t global_data[100];
> 
> int main(void) {
> 
>     str2_t var_data;
>     str2_t *var_data_p = &var_data;
>     uint64_t var_offset[100]; // = global_data; //= 0xdeadbeef;
>     
>     printf("size=%d\n", sizeof(str1_t)/sizeof(uint64_t));
> 
>     str1_t  tmp;
>     uint64_t *uint_p = (uint64_t *)&tmp;
>     
>     uint_p[0] = 5; // ESBMC fails on this line.
>     
>     uint_p[1] = 5;
>  
>     for (uint16_t i=1; i < sizeof(str1_t)/sizeof(uint64_t); i++) { // if the for statement starts fom zero ESBMC fails.
>         //((uint64_t *)&var_data_p->field1)[i] = 5; // ((uint64_t *)var_offset)[i];
>         uint_p[i] = 5;
>     }
>    return 0; 
> }
> ```

@lucasccordeiro , for this example ESBMC reports:
```
Counterexample:
size=8


State 2 file bitfields-7.c line 34 function main thread 0
----------------------------------------------------
Violated property:
  file bitfields-7.c line 34 function main
  dereference failure: Over-sized read of struct field


VERIFICATION FAILED
``` 
but in my opinion there is noting wrong with the code so it should be `VERIFICATION SUCCESSFUL`. I think this is caused by the fact that ESBMC treats access to the multiple struct members as a memory safety issue. But I think this should be considered safe (from the memory safety point of view) as long as this does not overflow the struct itself. So instead of reporting an error we could output a warning, or make it optional through a flag (e.g., `--fail-on-oversized-struct-field-access`).

lucasccordeiro commented:
@rafaelsamenezes: do you have some ideas of how to print these counterexamples nicely? I remember that you have done something similar to Test-Comp.

```
State 6 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 | CONCAT(byte_extract_little_endian({ 12, 0, 0, 0, 0, 0, 0, 0 }, 1), byte_extract_little_endian({ 12, 0, 0, 0, 0, 0, 0, 0 }, 0)) & 65039

```

rafaelsamenezes commented:
> @rafaelsamenezes: do you have some ideas of how to print these counterexamples nicely? I remember that you have done something similar to Test-Comp.

I did it in the wrapper. However, I will probably have to look into this again for witness validation.

mikhailramalho commented:
> @mikhailramalho , I've removed the unnecessary simplifications in the `symex_assign_bitfield` now. As for `compute_pointer_offset` we still need to keep it as it's used elsewhere in the code.
> 
> Regarding the counter-examples, I've added a testcase (`esbmc/regression/esbmc/struct_bitfields_15/`) which produces the following counter-example:
> 

It's weird to see that all those assignments happen in a single line (line 19), this is probably due to esbmc deconstructing the three steps you've created.

Although I don't think it should prevent the PR from being merged, we need to improve that and create a single ssa assignment, both for clarity and for the sake of the witness validators. Could you create an issue to follow-up on that?

Another solution would be to improve our simplifier to deal with expressions like:
```
State 3 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 | 0 & -2

State 4 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 0 | 0 & -3

State 5 file main.c line 19 function main thread 0
----------------------------------------------------
  *ptr = 12 | 0 & -13
```
at symex time, then rely on constant folding to improve the readability of the cex.

shmarovfedor commented:
I've tried this PR on the MemSafety category of SV-COMP and compared it against master branch. These are the results I got:

bitfields-fix:
```
Statistics:           3437 Files
  correct:            2419
    correct true:     1374
    correct false:    1045
  incorrect:             3
    incorrect true:      3
    incorrect false:     0
  unknown:            1015
  Score:              3697 (max: 5233)
```

master:
```
Statistics:           3437 Files
  correct:            2424
    correct true:     1380
    correct false:    1044
  incorrect:             7
    incorrect true:      3
    incorrect false:     4
  unknown:            1006
  Score:              3644 (max: 5233)
```

The total score has slightly gone up (by 53 points) as we managed to reduce the number of incorrect verdicts from 7 to 3 (which was pretty much expected as the main purpose of this PR was to fix these 4 cases). At the same time the number of unknowns is lower for master branch (1006 vs 1015). However, it seems that this number varies between different runs even for the same binary. So it's quite hard to judge from these experiments whether this PR introduces any significant overhead.

Nevertheless, I will try running this PR on the entire SV-COMP dataset this week to get a bigger picture.

shmarovfedor committed:
[migrate] fixed union flattening (struct case)

shmarovfedor committed:
[regression] added testcase for union flattening

shmarovfedor committed:
Merge branch 'master' into bitfields-fix

shmarovfedor committed:
Merge branch 'bitfields-fix' of github.com:shmarovfedor/esbmc into bitfields-fix

shmarovfedor commented:
I've fixed the issue pointed out by @fbrausse in test6.c. Also I have this PR running on the entire SV-COMP dataset, and the results should be available shortly. Meanwhile, I will look into the cex simplification problem.

lucasccordeiro commented:
Excellent, @shmarovfedor. Well done! 

Try speaking with @rafaelsamenezes about the algorithm to handle bytes extract for little- and big-endian. He has implemented a solution in Python, which could be converted to our code base.

mikhailramalho reviewed (approved):
I'm fine with the patch as long as we didn't regress any test on sv-comp.

shmarovfedor commented:
This is what I got from running this PR on the full SV-COMP:
```
Statistics:          15457 Files
  correct:            7470
    correct true:     5006
    correct false:    2464
  incorrect:            40
    incorrect true:     33
    incorrect false:     7
  unknown:            7947
  Score:             11308 (max: 25029)
```

The master branch is still running. I will post the results over the weekend.

lucasccordeiro commented:
Thanks, @shmarovfedor. Last time I run the master branch with the same parameters in April 30th, 2021, I obtained the following results (cf. https://github.com/esbmc/esbmc/issues/437):

```
Statistics:          15201 Files
  correct:            7717
    correct true:     4906
    correct false:    2811
  incorrect:            58
    incorrect true:     25
    incorrect false:    33
  unknown:            7426
  Score:             11295 (max: 24562)
```

Overall, we have done good progress since April 2021. However, the number of correct false results also dropped significantly.

mikhailramalho commented:
Hopefully it won't be that different (I see about 200 more test cases).

BTW, do you know what's failing in these 40 cases? I guess there should be
a lot of juliet.

On Thu, 11 Nov 2021, 13:07 Lucas Cordeiro, ***@***.***> wrote:

> Thanks, @shmarovfedor <https://github.com/shmarovfedor>. Last time I run
> the master branch with the same parameters in April 30th, 2021, I obtained
> the following results (cf. #437
> <https://github.com/esbmc/esbmc/issues/437>):
>
> Statistics:          15201 Files
>   correct:            7717
>     correct true:     4906
>     correct false:    2811
>   incorrect:            58
>     incorrect true:     25
>     incorrect false:    33
>   unknown:            7426
>   Score:             11295 (max: 24562)
>
> Overall, we have done good progress since April 2021. However, the number
> of correct false results also dropped significantly.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/482#issuecomment-966469977>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH3IXEZXNM3N5DDVO4DULPZ3TANCNFSM5BVYAP7A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

shmarovfedor commented:
These are the results from running master branch on the entire SV-COMP dataset.
```
Statistics:          15457 Files
  correct:            7476
    correct true:     5015
    correct false:    2461
  incorrect:            46
    incorrect true:     33
    incorrect false:    13
  unknown:            7935
  Score:             11227 (max: 25029)
```
Looks like overall we gain about 80 points by applying this PR. Despite having slightly more unknowns (7935 in master vs 7947 in this PR), this PR resolves 6 more incorrect verdicts (and we were aware of only 4 of them before).

@lucasccordeiro , I'm not sure we can directly compare these runs to any previous analysis, as for these two runs I compiled executables on the same machine, using the same flags, etc., and that version of ESBMC from April 2021 was produced on a different machine (hence, the number of unknowns can differ significantly).

lucasccordeiro commented:
That's great.

Since we keep our regression and fix some incorrect results, we have agreed to merge this PR into the master branch.

lucasccordeiro commented:
@shmarovfedor: could please rebase this branch with the master branch before we merge your PR?

shmarovfedor committed:
Merge branch 'master' into bitfields-fix
